Add config module as per #2349

belluzj opened the pull request:
See issue #2349

belluzj commented:
I tried to move the `register_option(...)` to be located in the module that is concerned with the option, but that creates an issue when parsing the configuration: if the module in question is not imported yet, its configuration options are not registered yet and so the contents of the configuration cannot be validated.

This could be solved by:
* listing all options in the config module instead of in the various modules
* deferring validation of config contents to the point where the option values are used, hoping that the code that uses the option is in the same module that registered the option, so the option will have been registered by then (I don't like this because I think it would be better for the user if they get a crash asap if the config contains errors)
* importing all modules that declare options when the config module is loaded (not great because why import code that you won't use?)
* having in each module a `config` file that registers options for the module, and import all of these when the top-level `config` module is loaded. Eg. `otlLib.config`, `varLib.config` etc

@anthrotype or @behdad  would you be able to advise please?

![image](https://user-images.githubusercontent.com/3616772/134660828-fc03651c-a163-4822-9fe2-b94a07748c27.png)

anthrotype commented:
The first option -- "listing all options in the config module instead of in the various modules" --  is the easiest, though probably not the most elegant.
The fourth option (otlLib.config, varLib.config etc) may also work, however in fonttools the subpackages `__init__.py` are usually not empty, so importing `varLib.config` would also import varLib's `__init__`.

What about, instead of needing to have calls to `register_option()` from some python module, you define a text file or files (e.g. configparser's .ini files) that contain the names of the modules and the options and the respective default values, and upon importing the global `fontTools.config` we parse that file or files (fetching it via importlib.resources and embedded in the distribution as package_data) and populate the Config class (calling register_option in turn for each one defined in the default config)?

belluzj commented:
> a text file or files (e.g. configparser's .ini files) that contain the names of the modules and the options and the respective default values

That sounds OK except it's not Python code, so one would not be able to implement parsing/validation logic (or one would have to express that declaratively, like with JSON Schema for example). For that reason I would say it would be better as Python code. Also, I was quite happy with the idea that you could import the config options into your code, to refer to them without having hardcoded strings in the code; that wouldn't be simple with the text-file-approach.

Maybe the first option can be made to look organized enough by having `config.varLib`, `config.otlLib` etc modules to store the options of each module? It's still the same as option 1 but maybe it would make immediate sense to contributors. What do you think about that?

anthrotype commented:
> implement parsing/validation logic

anthrotype commented:
> implement parsing/validation logic

can you make an example of what you have in mind? I thought the options values would be simple str, float, int, or stuff that you can store as json.

> you could import the config options into your code, to refer to them without having hardcoded strings in the code

Not sure what you mean by hard coded string. You could still import the default config object and get the already-parsed values from it.

anthrotype commented:
> config.varLib, config.otlLib etc modules to store the options of each module?

yeah, if you actually need to have _code_ to register options, then that could be an option to keep it organized. I'm still not 100 % that a declarative approach wouldn't be sufficient for our use case.

belluzj commented:
> can you make an example of what you have in mind? I thought the options values would be simple str, float, int, or stuff that you can store as json.

Yes it's indeed very simple stuff, but still, assuming people can give options by writing them in a config file, we should expect to get the wrong type or values out of range. JSON schema would handle these cases for sure. See the tests below: parsing an input string to int, checking that the value is in the expected range.

https://github.com/fonttools/fonttools/blob/bd0e1563ead0f309237cff15800e8a169caf63fb/Tests/config_test.py#L44-L58

> Not sure what you mean by hard coded string. You could still import the default config object and get the already-parsed values from it.

See the two tests below: you have the `config` object, and you can index it either by string or by the `Option` object. I feel that indexing by option object should be the preferred way, it would help prevent typos; but I admit it's not a massive win at all, just a small thing.

https://github.com/fonttools/fonttools/blob/bd0e1563ead0f309237cff15800e8a169caf63fb/Tests/config_test.py#L20-L30

behdad commented:
Collecting options from source into a toplevel config.py module is another option. We can make that run in CI and fail CI if config.py changed.

behdad commented:
I'll comment on the rest of the approach soon. Thanks!

belluzj commented:
Cosimo: don't perform conversions from string to int in the default constructor; have a special one for parsing from file that does the conversions. In general you wouldn't expect an object that accepts an int value to also accept a string value and convert it.

belluzj commented on Lib/fontTools/otlLib/optimize/gpos.py:
```diff
@@ -3,22 +3,42 @@
+from fontTools.config import register_option
+
+MODE_OPTION = register_option(
+    name="otlLib.optimize.gpos.mode",
```
Cosimo: prepend `fontTools.`, actually use `f'{__name__}.mode'` (but that only works if the option is defined in the module)

belluzj replied:
@anthrotype I did this but I went with the simple option 1, so I'm not using `__name__`

belluzj commented:
Instead of reading configuration from files, we could read it from a lib key in the designspace / UFO or a custom parameter in Glyphs.app source. 

E.g. in designspace:
```xml
<lib>
  <key>org.fontTools.config</key>
  <dict>
    <key>fontTools.otlLib.optimize.gpos.mode</key>
    <int>8</int>
  </dict>
</lib>
```

in ufo2ft: 

```python
ttFont = TTFont(config=designspace.lib.get('org.fontTools.config', {}))
```

belluzj commented:
Cosimo: should this module be in misc? It's not font-related

Also maybe there's an off-the-shelf thing in general for all this?

behdad commented:
> Instead of reading configuration from files, we could read it from a lib key in the designspace / UFO or a custom parameter in Glyphs.app source.

It can come from there as well, yes.



> ```
>     <key>fontTools.otlLib.optimize.gpos.mode</key>
> ```

I find `mode` very confusing. Can we call this `compressionLevel`?

behdad commented:
I feel like the module is fundamental enough that should be at toplevel `fonttools.config`. But I'm fine if it goes in `misc.config`, and for everyone to use. That's a pattern we've been following. And I don't see anything simple off the shelf that we can use.

As for the TTFont property, at the risk of bikeshedding, I suggest we call it `font.cfg`. :D.

belluzj commented on Lib/fontTools/varLib/plot.py:
```diff
@@ -2,7 +2,7 @@
 from fontTools.varLib.models import VariationModel, supportScalar
 from fontTools.designspaceLib import DesignSpaceDocument
-from matplotlib import pyplot
+# from matplotlib import pyplot
```
TODO: undo this change

belluzj commented:
* about where to define options: I implemented the simplest solution (all options to be listed in the config module's `__init__.py`)
* the module stays top-level but it's split between `fontTools.config.generic` (not related to fonts) and `fontTools.config` (the one to use everywhere in fontTools and client libraries). I could move the `generic` part to `misc` if that makes sense? Or not.
* I renamed `TTFont.config` to `TTFont.cfg` and `mode` to `compressionLevel`
* I'm using a dataclass in the config module and rebased on top of `drop-3.6` so that tests pass (sorry for the noise in the diffs, I'll rebase on top of main when possible)

belluzj commented:
@anthrotype @behdad if you agree with how it is at the moment I'll start adding other options. I would also be happy to know which options you have in mind to add; I'm think to start with the parameters to TTFont().

anthrotype commented on Lib/fontTools/config/generic.py:
```diff
@@ -0,0 +1,190 @@
+
+        self._values[name] = value
+
+    def get(self, option: Union[Option, str], default=_USE_GLOBAL_DEFAULT):
```
what's the point of allowing either an Option or a str for the key parameter of .get() and other dict-like methods? you're only using the Option's `name` attribute, a str, to lookup the global OPTIONS dict, so I think just accepting str would be enough, no?
What's the rationale?

belluzj replied:
My idea here was that it is (marginally) more convenient and prevents typos. Imagine with just strings:

```python
def function():
    font.cfg['long option name without typo']
    ...lots of code...
    if something:
        font.cfg['long option name with a typo']
    ...lots of code...
    font.cfg['long option name without typo']
```
* long strings everywhere
* no editor autocomplete when using options/no squiggly underlines from the linter
* the typo would be silent until the code reaches the line of code with the typo (hopefully there's a unit test that covers it)

with the PR as it is now:

```python
THING = OPTIONS['long option name without typo']

def function():
    font.cfg[THING]
    ...lots of code...
    if something:
        font.cfg[THING]
    ...lots of code...
    font.cfg[THING]
```

* long string only once at the top
* autocomplete + linter to pick typos in the constant name
* if there's a typo in the long string it will explode on import

I do agree that these are subjective and marginal benefits, I'm happy to simplify the code and just use strings.

anthrotype replied:
well, one could still do `font.cfg[THING.name]`.

behdad commented:
Thanks Jany. This is concise and good. I have some ideas forward. Let me sketch those.

belluzj commented on Lib/fontTools/config/__init__.py:
```diff
@@ -0,0 +1,36 @@
+"""
+from textwrap import dedent
+
+from fontTools.config.generic import *
```
Yes that makes sense, will do! I forgot about the __init__ behaviour you mention.

belluzj commented on Lib/fontTools/otlLib/optimize/gpos.py:
```diff
@@ -3,40 +3,16 @@
 
 log = logging.getLogger("fontTools.otlLib.optimize.gpos")
 
+LEVEL_OPTION = OPTIONS['fontTools.otlLib.optimize.gpos.compressionLevel']
```
Will do both. The colon sounds a bit strange to me at first glance but maybe it will make lots of sense once I see it in the code.

behdad commented:
Copy/pasting from chat to Jany/Cosimo:

Jany, Cosimo, I have some Evil config ideass 😀
gimme a week to cook them.
Here's the sneak peak:
based on your design so far.
In your __getitem__
```
def __getitem__(s):

  if hasattr(s, '__module__'):
    s = s.__module__ + s.__name__
  elif hasattr(s, '__name__'):
    s = s.__name__
```
then accept a tuple: (path,node):
```
  s = path+node
```
in short, you can then do:
```
  cfg[obj,"COMPRESSION_LEVEL"]
```
and avoid repeating `"fontTools.ttLib....."`

any classtype or module can be passed there.

better yet, instead of a tuple, we can accept a colon as Cosimo liked 😛
```
  cfg[obj:"COMPRESSION_LEVEL"]
```

Next. If only a path is passed and no node, we return the subconfig. That is, `cfg[cls]` will return a new cfg that has subconfig for the class. `__getattr__` will also be overriden to do the right thing.  So we can do: `cfg[cls].COMPRESSION_LEVEL`

Finally, `__call__` will be used to set config on multiple objects as a context:

```
with cfg(obj1, obj2, obj3, mod1, cls1, ...):
  ...
```
and the main way to use the config is to use to populate key/value arguments of constructors, methods, and functions. The syntax for that I have not fully figured out yet.
```
  cfg.invoke(obj.method) (params)
```
That would pass all relevant cfg to method.

Anyway; this is what's been brewing in my head.

As for your Option type, I want use the typing module types directly.

No need for extra Option type.

Behdad Esfahbod,
Lemme copy/paste this on the issue.

Jany Belluz,
Ok, thanks

That sounds a bit complicated to be honest, but maybe it looks nice when using it in the code

Behdad Esfahbod,
Yeah I feel like ergonomically it works.

Jany Belluz,
Ok cool

(sounds like how a Ruby developer would think, not a Python one)
I was going for pedestrian and non-surprising

Behdad Esfahbod,
I understand 🙂

I'm not sure mine are good. Evil it might be.

But want to explore 😀

Jany Belluz,
Yes anyway on the issue would be good 🙂 I'm curious to see how the usage looks

belluzj commented:
@anthrotype I rebased on latest main

anthrotype commented on Lib/fontTools/misc/configTools.py:
```diff
@@ -0,0 +1,186 @@
+_USE_GLOBAL_DEFAULT = object()
+
+
+class Config(MutableMapping):
```
I suggest we give Config a nice `__repr__` to aid debugging, I think MutableMapping doesn't make one for you.

belluzj replied:
Print `f"Config({repr(self._values)})"`

anthrotype commented on Lib/fontTools/misc/configTools.py:
```diff
@@ -0,0 +1,186 @@
+    help: str
+    default: Any
+    # Turn input (e.g. string) into proper type. Only when reading from file.
+    parse: Callable
```
Callable[[str], Any]

anthrotype commented on Lib/fontTools/misc/configTools.py:
```diff
@@ -0,0 +1,186 @@
+    # Turn input (e.g. string) into proper type. Only when reading from file.
+    parse: Callable
+    # Return true if the given value is an acceptable value
+    validate: Callable
```
Callable[[Any], bool]

anthrotype commented on Lib/fontTools/misc/configTools.py:
```diff
@@ -0,0 +1,186 @@
+        try:
+            option = OPTIONS[name]
+        except KeyError:
+            # FIXME: (Jany)  What to do with options we don't recognize?
```
could be an init parameter to the Config class, like `skip_unknown=False`

belluzj replied:
Cosimo says: Or maybe a version the config file? Needs more thinking (maybe see what other applications are doing with their config). Maybe set() should always raise if you try to set an unknown option, but when you're parsing a file, maybe there it would make sense to skip unknown options?

belluzj replied:
I added skip_unknown=False

belluzj commented on Lib/fontTools/misc/configTools.py:
```diff
@@ -0,0 +1,186 @@
+
+@dataclass
+class Option:
+    name: str
```
Cosimo: delete this attribute, not needed because it's already in the dictionary key of OPTIONS

belluzj commented on Lib/fontTools/misc/configTools.py:
```diff
@@ -0,0 +1,186 @@
+    validate: Callable
+
+
+OPTIONS: Dict[str, Option] = {}
```
Make object with Mapping interface, have register_option as method, move global variable to fontTools.config.__init__ (and have the inner dict of options double underscore prefixed)

belluzj commented on Lib/fontTools/misc/configTools.py:
```diff
@@ -0,0 +1,186 @@
+
+    @classmethod
+    def from_file(cls):
+        """TODO"""
```
I removed the method for now

belluzj commented on Lib/fontTools/varLib/merger.py:
```diff
@@ -850,10 +846,10 @@ def merge(merger, self, lst):
 		# smaller subtables, which may prevent overflows from happening.
-		mode = os.environ.get(GPOS_COMPACT_MODE_ENV_KEY, GPOS_COMPACT_MODE_DEFAULT)
-		if mode and mode != "0":
+		level = merger.font.cfg[COMPRESSION_LEVEL.name]
```
I removed that and used the string directly everywhere

belluzj commented:
@anthrotype I fixed all your comments and added the new module to the Sphinx documentation.

I had an idea: how neat would it be if the list of registered options was parsed by Sphinx and added to the documentation page as well? I had a quick look and it sounds possible by adding some Sphinx code with new nodes and stuff, but I didn't go down the rabbit hole. Just a thought for later :)

anthrotype commented:
> would it be if the list of registered options was parsed by Sphinx 

that would be nice, yeah. File an issue so we don't forget

belluzj commented:
Done for the classvar and the issue

anthrotype commented on Lib/fontTools/ttLib/ttFont.py:
```diff
@@ -101,6 +102,7 @@ def __init__(self, file=None, res_name_or_index=None,
 		self.recalcTimestamp = recalcTimestamp
 		self.tables = {}
 		self.reader = None
+		self.cfg = Config(cfg)
```
how about here you check `if not isinstance(cfg, AbstractConfig): cfg = Config(cfg)` otherwise you self `self.cfg = cfg`. This would allow one to stick a custom super-set of fonttools own Config with additional flags that can be useful in ones workflow

belluzj replied:
Ok will do now

anthrotype replied:
We would still want each TTFont instance to have its distinct copy, since a config is mutable.
You current cast `Config(cfg)` achieves that already. To accomplish the same with my proposal above we need a `AbstractConfig.copy()` method that returns a new instance of the type with its own copy of the internal `self._values` dictionary. A shallow copy is fine, if any mutable values are used in individual configs, it's not our job to deepcopy those I think.

So in addition to [my earlier suggestion](https://github.com/fonttools/fonttools/pull/2416/files#r850336161), we add this to `AbstractConfig`:

```python
def copy(self):
    return self.__class__(self._values)
```

and then call it inside `TTFont.__init__`

```python
    self.cfg = Config(cfg) if not isinstance(cfg, AbstractConfig) else cfg.copy()
```

anthrotype replied:
maybe also add a test that sets identical Config oject in two different fonts, and assert they have different `id()`;  or modify one font.cfg, and assert that it did not also change in the other font

belluzj replied:
Done, I added a test for the use-case that you were describing, if you can check that it's what you meant?

anthrotype replied:
> if you can check that it's what you meant?

yes, that's it

belluzj replied:
Done for real now including the copy()

anthrotype commented on Lib/fontTools/ttLib/ttFont.py:
```diff
@@ -89,7 +91,7 @@ def __init__(self, file=None, res_name_or_index=None,
 			verbose=None, recalcBBoxes=True, allowVID=NotImplemented, ignoreDecompileErrors=False,
 			recalcTimestamp=True, fontNumber=-1, lazy=None, quiet=None,
-			_tableCache=None):
+			_tableCache=None, cfg={}):
```
in general mutable default arguments are bad, but in this case you always make a copy so it should be fine. Alternatively you default to `cfg=None` and then in method body do:

```
if cfg is None:
    cfg = Config()
self.cfg = cfg.copy() if isinstance(cfg, AbstractConfig) else Config(cfg)
```

anthrotype replied:
actually it's fine like this

belluzj commented:
@anthrotype I wrote release notes in the description
